### PR TITLE
Make s3 upload sink PUT small objects

### PR DESCRIPTION
**Pull request by xemul**
When upload-sink is flushed, it may notice that the upload had not yet been started and fall-back to plain PUT in that case. This will make small files uploading much nicer, because multipart upload would take 3 API calls (start, part, complete) in this case

fixes: #13014

### CODE REVIEW

File: test/boost/s3_test.cc [hunk 1]
```diff
@@ -162,6 +162,34 @@ SEASTAR_THREAD_TEST_CASE(test_client_multipart_copy_upload) {
...
+    auto close = seastar::deferred_close(out);
+
+    temporary_buffer<char> data = sstring("1A3B5C7890").release();
+    out.write(reinterpret_cast<const char*>(data.begin()), data.size()).get();
+
+    testlog.info("Flush upload");
+    out.flush().get(); // if it tries to do regular flush, memory claim would throw
+    auto delete_object = deferred_delete_object(cln, name);
+
+    testlog.info("Closingn");
```
Comment (tchaikov) [inline test/boost/s3_test.cc:185]: might want to drop the "n" postfix? if it does not serve some certain purpose which i am unaware of.

Comment (xemul) [reply test/boost/s3_test.cc:185]: Ah :) It's a `\n` not completely removed. Thanks

Comment (xemul) [reply test/boost/s3_test.cc]: Fixed

**Comment by xemul** [comment]
upd:
- logging messages tuned

**Review by tchaikov — APPROVED** [review]
thank you. i consider this change a performance improvement.

**Comment by xemul** [comment]
#15832

**Pull request merged**